bottomless/src/replicator.rs

```diff
@@ -616,23 +618,17 @@ impl Replicator {
                     .body(ByteStream::from_path(&self.db_path).await?)
                     .send()
                     .await?;
-                let mut reader = tokio::fs::File::open(&self.db_path).await?;
-                Self::read_change_counter(&mut reader).await?
             }
             CompressionKind::Gzip => {
-                // TODO: find a way to compress ByteStream on the fly instead of creating
-                // an intermediary file.
-                let (compressed_db_path, change_counter) = self.compress_main_db_file().await?;
+                let buf_reader = tokio::io::BufReader::new(reader.try_clone().await?);
+                let gzip_reader = async_compression::tokio::bufread::GzipEncoder::new(buf_reader);
+
                 let key = format!("{}-{}/db.gz", self.db_name, self.generation);
-                self.client
-                    .put_object()
-                    .bucket(&self.bucket)
-                    .key(key)
-                    .body(ByteStream::from_path(compressed_db_path).await?)
-                    .send()
-                    .await?;
-                let _ = tokio::fs::remove_file(compressed_db_path).await;
-                change_counter
+
+                // Unfortunally we can't send the gzip output in a single call without buffering
+                // the whole snapshot in memory because S3 requires the `Content-Length` header
+                // to be set.
```

Comment:
It seems that this comment is a bit misleading. Even Gzipped, snapshot may take GBs in size, which is absolute no go to be kept as a whole in memory.

Reply:
What about:
```
Since it's not possible to know the exact size of a gzip stream and the `PutObject` operation requires the `Content-Length` header to be set, we need to send the content in chunks of known size.
```